torch/masked/maskedtensor/core.py

```diff
@@ -0,0 +1,533 @@
+    def backward(ctx, grad_output):
+        (layout_tensor,) = ctx.saved_tensors
+        layout = INT_TO_LAYOUT[layout_tensor.item()]
+
+        if layout == torch.sparse_coo:
+            return grad_output.to_sparse_coo()
+        elif layout == torch.sparse_csr:
+            return grad_output.to_sparse_csr()
+        elif layout == torch.strided:
+            return grad_output.to_dense()
+        raise ValueError("to_dense: Unsupported input layout: ", layout)
+
+
+class MaskedToSparse(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input):
+        assert is_masked_tensor(input)
+        # Following the convention from sparse  tensors that to_sparse always means that we convert to sparse_coo
+        if input.layout() == torch.sparse_coo:
+            return input
+
+        mask = get_mask(input)
+        data = get_data(input)
+
+        sparse_mask = mask.to_sparse_coo().coalesce()
+        sparse_data = data.sparse_mask(sparse_mask)
+
+        return MaskedTensor(sparse_data, sparse_mask)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output.to_dense()
+
+
+class MaskedToSparseCsr(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input):
+        assert is_masked_tensor(input)
+        assert (
+            input.masked_data.ndim == 2
+        ), f"Only 2D tensors can be converted to the SparseCsr layout but got shape: {input.masked_data.size()}"
+
+        if input.layout() == torch.sparse_csr:
+            return input
+
+        mask = get_mask(input)
+        data = get_data(input)
+
+        sparse_mask = mask.to_sparse_csr()
+        sparse_data = data.sparse_mask(sparse_mask)
+
+        return MaskedTensor(sparse_data, sparse_mask)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output.to_dense()
+
+
+# Needed until https://github.com/pytorch/pytorch/issues/65243 is fixed
+# since derivative includes usage of zeros_like
+# https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L1516-L1519
+class MaskedWhere(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, cond, self, other):
+        # logger.debug(
+        #     "Calling MaskedWhere.forward")
+        ctx.mark_non_differentiable(cond)
+        ctx.save_for_backward(cond)
+        return torch.ops.aten.where(cond, self, other)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        # logger.debug("Calling MaskedWhere.backward")
+        (cond,) = ctx.saved_tensors
+
+        def masked_out_like(mt):
+            return MaskedTensor(get_data(mt), torch.zeros_like(get_mask(mt)).bool())
+
+        return (
+            None,
+            torch.ops.aten.where(cond, grad_output, masked_out_like(grad_output)),
+            torch.ops.aten.where(cond, masked_out_like(grad_output), grad_output),
+        )
+
+
+class MaskedTensor(torch.Tensor):
+    @staticmethod
+    def __new__(cls, data, mask, requires_grad=False):
```

Comment:
Should you check that data and mask are both plain Tensors here? Or at least that data is not  MaskedTensor